enhancements/microshift/etcd-as-transient-systemd-unit.md

```diff
@@ -0,0 +1,123 @@
+last-updated: yyyy-mm-dd
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - TBD
+see-also:
+  - "/enhancements/this-other-neat-thing.md"
+replaces:
+  - "/enhancements/that-less-than-great-idea.md"
+superseded-by:
+  - "/enhancements/our-past-effort.md"
+---
+
+# etcd as a Transient systemd Unit
+
+## Summary
+
+The enhancement proposes moving etcd out of the MicroShift binary and into a transient systemd unit that is launched and managed by MicroShift.
+
+## Motivation
+
+Currently MicroShift [bundles upstream etcd](https://github.com/openshift/microshift/blob/f7f2260e4fbff61654f478fc149cb7052261f87f/go.mod#L21) into its binary and runs it as a goroutine. This results in etcd dependencies being coupled with MicroShift which makes it harder to maintain and upgrade etcd independently of the platform like it is done with a separate openshift/etcd repo in the Openshift Container Platform (OCP).
+
+This enhancement outlines an update to MicroShift that would decouple its deployment of etcd to bring it closer to how etcd is maintained with OCP.
+
+Changing the deployment of etcd, outlined below, would alleviate concerns about building etcd with a version of shared dependencies - mainly library-go - that etcd is not currently built against, and therefore, not tested against.
+
+Additionally moving the etcd deployment into its own process would make it easier to debug and process logs via journatlctl.
+
+### User Stories
+
+1. "As a developer, I want to maintain etcd versions independently of microshift dependencies."
+
+2. "As an admin, I want to observe and debug the etcd server logs separately from other microshift processes."
+
+### Goals
+
+- The etcd binary is built and shipped separately from the microshift binary to ease maintainence
+- The etcd server logs are easier to observe
+
+### Non-Goals
+
+- 
+
+## Proposal
+    
+#### Running etcd as a transient systemd unit
+In order to address the concerns brought up in the Motivation section, the MicroShift binary will need to be updated to remove the currently bundled etcd server and replaced with a forked process running under systemd; using the `systemd-run` command. Only the execution of etcd will change - from goroutine to external process - MicroShift's general usage of etcd will not be changed.
+    
+Additionally, this external process will need to be shutdown when MicroShift exits. In an ungraceful termination of MicroShift, it is possible that the etcd systemd unit will continue execution, so it may be necessary to add code to detect this on MicroShift startup so that another etcd launch is not attempted: this would be a port conflict and cause the second etcd instance to fail to start. Further invesigation is needed to learn if a transient systemd unit's lifecycle can be tied to another non-transient systemd unit (MicroShift) so that if the latter exits ungracefully, the former will be gracefully shutdown.
```

Comment:
Leaving etcd running when MicroShift is stopped is going to give us a lot of other orchestration issues. Is there no way to use systemd to ensure that the ephemeral unit stops when the MicroShift process is not present?

Reply:
Why would we need a transient systemd unit?, can't we simply define an etcd system unit and make microshift depend into the etcd unit? I believe this is more desirable than managing systemd units from microshift (maintaining the OS, which is something we don't want to do from MicroShift)

What would be the risk of leaving etcd running when MicroShift is stopped?

I propose we create etcd's own systemd unit and use the standard dependencies with systemd.

Reply:
@dhellmann Not to my knowledge, I haven't yet found an argument to `systemd-run` that keeps the lifetime of the 'subprocess' bound to the lifetime of the main process. Currently, this acts like a 'start', and it will need to be cleaned up later, with `systemctl stop microshift-etcd`.

@mangelajo The reasoning behind using a transient unit was so that MicroShift could start and configure before launching etcd (for generating certs and such). I think that keeping the execution lifetime of etcd managed by MicroShift is the right way to go for that reason, but I defer to Doug if he thinks that with a normal systemd unit we could still get the pre-launch configuration.

Reply:
@mangelajo the transient unit came in because we were trying to stick to the architectural pattern of the user only having to start and manage one process. If we have an RPM that installs a separate unit, we lose that ability.

Reply:
I want to emphasize the avoidance of orchestration issues and to make this transparent for the users. The UX should be a high priority for the project.

Reply:
Generation of certificates seems like a good reason for MicroShift to control etcd via transient systemd unit, or probably better a regular unit that MicroShift starts/stops and provides certificates for

Reply:
We have 2 runtime modes in MicroShift: systemd (production) and command line (development). If we're starting etcd process by forking it, the right mode should be detected and systemd-run should only be used when appropriate.

Reply:
@dusk125 Does this work from command line when in development mode? These instructions should be working after the etcd change.
https://github.com/openshift/microshift/blob/main/docs/devenv_rhel8.md#running-microshift

Reply:
I think Giga's point is also valid. Currently we can just `make` and run microshift binary.
This section assumes that MicroShift is always started as a systemd service which isn't a case for developer environments (and I wouldn't want to have that ability taken away).

Was managing etcd execution by microshift was already discussed (starting it from the code)?
Let say we take a route of etcd submodule within microshift repo. Built RPM would install etcd into standard location (like /usr/bin/etcd) so microshift would know where to look. If we'd want to rebuild etcd binary (question here is how often would that be, perhaps every rebase? that could be on daily basis...) then it'd require copying that produced binary into a "standard location". Or perhaps detect whether microshift is started by systemd or not (then look for etcd binary in the same dir as microshift's bin).

Reply:
I see a couple of ways to address the developer use case. There may be others.

If we put the 2 binaries in the same directory, then we could have microshift look for etcd in the directory based on the binary it is running from (rather than using a hard-coded path). That would allow the developer workflow to continue to rely on building and running from an arbitrary source directory (assuming we go ahead with the idea of creating a separate go module and continue vendoring etcd into the microshift repo).

If we don't like looking for the etcd binary dynamically, we could use a command line flag to override the etcd binary path and add a Makefile target in openshift/microshift for running the development version.

Reply:
I agree, spawning etcd as a child process of MicroShift (via os/exec) may be a better solution than handling it via a (transient) systemd unit. It'd give us the possibility to signal it from MicroShift and it ensures their lifecycles are tied together, too.

Reply:
Oh, no, I wasn't proposing dropping the systemd aspect. We want that to capture logs. I just meant that the path we give to systemd for the etcd binary could be dynamic.

Reply:
https://github.com/openshift/microshift/pull/976 [proposes](https://github.com/openshift/microshift/pull/976/files#diff-6d2df5f6740fc999909995be07371e8eb25745795233d4c4cbfe2fa6b1a891e8R77) using [BindsTo=](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#BindsTo=) - it looks like it allows only for service to service binding, not executable.
If that's the case: should we have alternative path for dev mode?

Reply:
Yes, we should have a dev-mode (local binary) vs systemd mode, that should be easily added when/after we add the transient systemd unit invocation